walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,94 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.function.Predicate;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<Event> eventsSortedByEnd = sortEvents(events, false);
+
+    clearIrrelevantEvents(eventsSortedByEnd, request.getAttendees());
+
+    return findAvailableTimeSlots(eventsSortedByEnd, request.getDuration());
+  }
+
+  private Collection<TimeRange> findAvailableTimeSlots(
+      Collection<Event> eventsSortedByEnd, long minDuration) {
+    Collection<TimeRange> options = new ArrayList<>();
+
+    if (eventsSortedByEnd.isEmpty()) {
+      options.add(TimeRange.WHOLE_DAY);
+    } else {
+      Event event = getFirstEvent(eventsSortedByEnd);
+      addTimeRangeIfPossible(options, TimeRange.START_OF_DAY, event.getWhen().start(), minDuration);
+
+      Iterator<Event> eventsIter = eventsSortedByEnd.iterator();
+      while (eventsIter.hasNext()) {
+        Event nextEvent = eventsIter.next();
+        if (!nextEvent.getWhen().overlaps(event.getWhen())) {
+          addTimeRangeIfPossible(
+              options, event.getWhen().end(), nextEvent.getWhen().start(), minDuration);
+        }
+        event = nextEvent;
+      }
+
+      addTimeRangeIfPossible(options, event.getWhen().end(), TimeRange.END_OF_DAY, minDuration);
+    }
+
+    if (TimeRange.WHOLE_DAY.duration() < minDuration) {
+      options.clear();
+    }
```

Comment:
You should probably move this if statement to be below the second if statement and add `return options` to the end of it

Reply:
Ok yea that would be cleaner.